include/libsemigroups/adapters.hpp

```diff
@@ -381,6 +382,62 @@ namespace libsemigroups {
     }
   };
 
+  ////////////////////////////////////////////////////////////////////////
+  // Hash specializations
+  ////////////////////////////////////////////////////////////////////////
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! This type is stateless trivially default constructible
+  //! and has a call operator of signature `size_t operator()(std::vector<T>
+  //! const&)` for use with, for example,
+  //! [std::unordered_map](https://en.cppreference.com/w/cpp/container/unordered_map).
+  //!
+  //! \tparam T the type of objects to compare in the vector.
+  //!
+  //! The second template parameter exists for SFINAE in overload resolution.
+  //!
+  //! \par Used by KoniecznyTraits.
+  template <typename T>
+  struct Hash<std::vector<T>> {
+    //! This call operator hashes \p vec.
+    //!
+    //! \param vec the value to hash
+    //!
+    //! \returns A hash value for \p x, a value of type `size_t`.
+    size_t operator()(std::vector<T> const& vec) const {
+      size_t val = 0;
+      for (T const& x : vec) {
+        val ^= Hash<T>()(x) + 0x9e3779b97f4a7c16 + (val << 6) + (val >> 2);
+      }
+      return val;
+    }
+  };
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! This type is stateless trivially default constructible
+  //! and has a call operator of signature `size_t operator()(std::vector<T>
+  //! const&)` for use with, for example,
+  //! [std::unordered_map](https://en.cppreference.com/w/cpp/container/unordered_map).
+  //!
+  //! \par Used by Konieczny.
+  template <>
+  struct Hash<std::pair<size_t, size_t>> {
+    //! This call operator hashes \p x.
+    //!
+    //! \param x the value to hash
+    //!
+    //! \returns A hash value for \p x, a value of type `size_t`.
+    size_t operator()(std::pair<size_t, size_t> const& x) const noexcept {
+#if LIBSEMIGROUPS_SIZEOF_VOID_P == 8
+      return (x.first << 32) + x.second;
```

Comment:
There was a comment about this being bad somewhere before, can we reinstate this comment?